src/python/Services/ReqMgr/ReqMgrReader.py

```diff
@@ -6,53 +6,383 @@
 
 import logging
 import os
+import time
+import copy
 from Utils.WebTools import getResponse
 from Utils.ConfigurationHandler import ConfigurationHandler
 
+from typing import List, Optional, Union, Any, Callable
+
 
 class ReqMgrReader(object):
     """
     _ReqMgrReader_
     General API for reading data from ReqMgr
     """
 
-    def __init__(self, url=None, logger=None, **contact):
-
+    def __init__(
+        self, url: str = None, logger: Optional[logging.Logger] = None, **contact
+    ) -> None:
+        # TODO: url is not being used
         try:
             configurationHandler = ConfigurationHandler()
-            self.reqmgrUrl = os.getenv('REQMGR_URL', configurationHandler.get('reqmgr_url'))
+            self.reqmgrUrl = os.getenv(
+                "REQMGR_URL", configurationHandler.get("reqmgr_url")
+            )
+            self.reqmgrEndpoint = {
+                "req": "/reqmgr2/data/request/",
+                "info": "/reqmgr2/data/info/",
+                "agentConfig": "/reqmgr2/data/wmagentconfig/",
+                "splitting": "/reqmgr2/data/splitting/",
+                "couchdb": "/couchdb/reqmgr_workload_cache/",
+            }
+
             logging.basicConfig(level=logging.INFO)
             self.logger = logger or logging.getLogger(self.__class__.__name__)
-        except Exception as e:
+
+        except Exception as error:
             msg = "Error initializing ReqMgrReader\n"
-            msg += "%s\n" % format(e)
+            msg += f"{error}\n"
             raise Exception(msg)
 
-    def getWorkflowByCampaign(self, campaign, details=False):
+    def _runWithRetries(
```

Comment:
This is a general comment: Let's drop `runWithRetries` in all functions. It really makes the code quite long and unreadable. Once the code goes into production, when we can enable this feature for unstable endpoints. Ideally, it would have been great if we can just list the functions for which you want to retry and all functions are run with retry policy when they are invoked. Anyhow, this is something to think about in the future. For now, let's drop it. We can still keep the function in the Utils package as I mentioned above in case it's needed.

Reply:
I was thinking of defining `runWithRetries` as a decorator, so it will be easier to use it in the future. What do you think?

Reply:
that makes sense to me.